gbpservice/nfp/tests/unit/nfp/configurator/test_firewall.py

```diff
@@ -0,0 +1,887 @@
+            context.update({'notification_data': {}})
+            context.update({'resource': request_data1['config'][0]['resource']})
+            del request_data1['config'][0]['kwargs']['context']
+            kwargs = request_data1['config'][0]['kwargs']
+            if bulk:
+                sa_info_list = self.fo.fake_sa_info_list()
+                if operation == 'delete':
+                    sa_info_list[0]['method'] = 'clear_interfaces'
+                    sa_info_list[1]['method'] = 'clear_routes'
+                args_dict = {
+                         'sa_info_list': sa_info_list,
+                         'notification_data': {}
+                        }
+            else:
+                args_dict = {'context': context,
+                             'kwargs': kwargs}
+            mock_sc_event.assert_called_with(id=method, data=args_dict)
+            mock_sc_rpc_event.assert_called_with('foo')
+
+    def _test_fw_event_creation(self, operation):
+        sc, conf, rpc_mgr = self._get_ConfiguratorRpcManager_object()
+        agent, sc = self._get_FWaasRpcManager_object(conf, sc)
+        arg_dict = {'context': self.fo.context,
+                    'firewall': self.fo._fake_firewall_obj(),
+                    'host': self.fo.host}
+        method = {'CREATE_FIREWALL': 'create_network_service_config',
+                  'UPDATE_FIREWALL': 'update_network_service_config',
+                  'DELETE_FIREWALL': 'delete_network_service_config'}
+        request_data = self.fo.fake_request_data_fw()
+        with mock.patch.object(sc, 'new_event', return_value='foo') as (
+                                                        mock_sc_event), \
+            mock.patch.object(sc, 'post_event') as mock_sc_rpc_event, \
+            mock.patch.object(rpc_mgr,
+                              '_get_service_agent_instance',
+                              return_value=agent):
+            getattr(rpc_mgr, method[operation])(self.fo.context, request_data)
+
+            mock_sc_event.assert_called_with(id=operation, data=arg_dict)
+            mock_sc_rpc_event.assert_called_with('foo')
+
+    def _test_notifications(self, bulk=False):
+        sc, conf, rpc_mgr = self._get_ConfiguratorRpcManager_object()
+        agent = self._get_GenericConfigEventHandler_object(sc, rpc_mgr)
+        
+        data = "PUT ME IN THE QUEUE!"
+        with mock.patch.object(sc, 'new_event', return_value='foo') as (
+                                                        mock_new_event), \
+            mock.patch.object(sc, 'poll_event') as (mock_poll_event):
+            
+            agent._notification(data)
+
+            mock_new_event.assert_called_with(id='NOTIFICATION_EVENT',
+                                              key='NOTIFICATION_EVENT',
+                                              data=data)
+            mock_poll_event.assert_called_with('foo')
+        
+    def test_configure_routes(self):
```

Comment:
All the following functions should be abstracted into single one. All the code is same except the method and operation. Lets simplify it.